Kernel/VM/MemoryManager.cpp

```diff
@@ -617,10 +617,66 @@ Region* MemoryManager::kernel_region_from_vaddr(VirtualAddress vaddr)
     return nullptr;
 }
 
+Region* MemoryManager::find_user_region_from_vaddr_no_lock(Space& space, VirtualAddress vaddr)
+{
+    VERIFY(space.get_lock().own_lock());
+    return space.find_region_containing({ vaddr, 1 });
+}
+
 Region* MemoryManager::find_user_region_from_vaddr(Space& space, VirtualAddress vaddr)
 {
     ScopedSpinLock lock(space.get_lock());
-    return space.find_region_containing({ vaddr, 1 });
+    return find_user_region_from_vaddr_no_lock(space, vaddr);
+}
+
+void MemoryManager::validate_syscall_preconditions(Space& space, RegisterState const& regs)
+{
+    // We take the space lock once here and then use the no_lock variants
+    // to avoid excessive spinlock recursion in this extemely common path.
+    ScopedSpinLock lock(space.get_lock());
```

Comment:
The comment seems to contradict with how the lock is held. It looks like the lock is held throughout the entire function?

Reply:
not sure what it contradicts? he holds the lock for the whole function, and then calls the no_lock variants inside it so they dont have to check the lock and lock it if needed

Reply:
Oh I see, got it. Somehow I assumed that the no_lock variants were able to run without the lock being held.